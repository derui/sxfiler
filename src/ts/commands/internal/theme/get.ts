import { Actions } from "@/modules";
import { CommandLike, CommandState, CommandDescriptor } from "@/commands/type";
import { Dispatcher } from "@/types";
import { actions } from "@/modules/theme";
import { GetRequest } from "@/generated/theme_pb";
import { Theme } from "@/rpc/client-procedures";

const identifier = "internal.theme.select";

export type Payload = undefined;
export type Command = CommandLike<Payload>;

/**
 * command descriptor to resolve
 */
export const descriptor: CommandDescriptor<Payload> = Object.freeze({
  identifier,
  payloadSpec: undefined as any,
});

export const createCommand = (): Command => {
  return {
    identifier,
    async execute(dispatcher: Dispatcher<Actions>, args: CommandState) {
      const res = await args.clientResolver.rpcClient().use(Theme.getTheme)(new GetRequest());

      const theme = res.getTheme();
      if (theme) {
        dispatcher.dispatch(actions.update(theme));
      }
    },
  };
};
