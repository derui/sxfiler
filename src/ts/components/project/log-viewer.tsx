import { h } from "preact";

export type Props = {
  entries: string[];
  hidden: boolean;
};

/**
 * make item for a log entry
 */
const makeItem = (entry: string) => {
  let level = "info";

  return (
    <div role="listitem" class="log-viewer__log-item" data-testid="logViewer-logItem" key={entry}>
      <span class="log-viewer__item-level" data-testid="logViewer-itemLevel" data-level={level}>
        [{level.toUpperCase()}]
      </span>
      <span class="log-viewer__item-message" data-testid="logViewer-itemMessage">
        {entry}
      </span>
    </div>
  );
};

export const Component: preact.FunctionComponent<Props> = ({ entries, hidden }) => {
  let components: preact.VNode<any>[] = [];
  if (!hidden) {
    components = entries.map(makeItem);
  }

  return (
    <div class="log-viewer__root" data-testid="logViewer-root" aria-hidden={hidden}>
      {components}
    </div>
  );
};
