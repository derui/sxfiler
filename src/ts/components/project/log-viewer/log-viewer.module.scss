@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.root {
  composes: list-base from "../../ui/list/list.module.scss";

  background-color: $sol_base03;
  color: $sol_base2;

  height: 100%;
  font-size: 1rem;
  overflow-x: auto;
  overflow-y: scroll;

  font-family: monospace;
}

.item {
  display: flex;
  flex-direction: row;
  padding: $nanoSpace 0;
}

.itemLevel {
  flex: 0 0 4rem;
  white-space: pre;
  margin-right: 1rem;

  &[data-level="info"]{
    color: $sol_blue;
  }

  &[data-level="warning"]{
    color: $sol_orange;
  }

  &[data-level="error"]{
    color: $sol_red;
  }
}

.itemMessage {
  flex: 0 0 auto;
}
