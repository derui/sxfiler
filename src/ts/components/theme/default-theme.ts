// styled-components.ts
import * as styledComponents from "styled-components";

export interface Theme {
  baseFontSize: string;
  baseBorderRadius: string;
  headerShadow: string;
  boxShadow: string;
  colors: {
    base03: string;
    base02: string;
    base01: string;
    base00: string;
    base0: string;
    base1: string;
    base2: string;
    base3: string;
    yellow: string;
    orange: string;
    red: string;
    magenta: string;
    violet: string;
    blue: string;
    cyan: string;
    green: string;
  };
  spaces: {
    large: string;
    base: string;
    nano: string;
    small: string;
  };
}

export const Theme = {
  baseFontSize: "13.5px",
  baseBorderRadius: "4px",
  headerShadow: "0px 7px 3px rgba(0, 0, 0, 0.2)",
  boxShadow: "4px 4px 7px rgba(0, 0, 0, 0.2), -4px 0px 7px rgba(0, 0, 0, 0.2)",

  colors: {
    /* color schema of solalized-dark (from http://ethanschoonover.com/solarized) */
    base03: "#002b36",
    base02: "#073642",
    base01: "#586e75",
    base00: "#657b83",
    base0: "#839496",
    base1: "#93a1a1",
    base2: "#eee8d5",
    base3: "#fdf6e3",
    yellow: "#b58900",
    orange: "#cb4b16",
    red: "#dc322f",
    magenta: "#d33682",
    violet: "#6c71c4",
    blue: "#268bd2",
    cyan: "#2aa198",
    green: "#859900",
  },
  spaces: {
    large: "16px",
    base: "8px",
    small: "4px",
    nano: "2px",
  },
};

const { default: styled, css, ThemeProvider } = styledComponents as styledComponents.ThemedStyledComponentsModule<
  Theme
>;

export { styled, css, ThemeProvider };
