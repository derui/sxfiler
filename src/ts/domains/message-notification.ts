export enum Level {
  Info = "info",
  Warning = "warning",
  Error = "error",
}

export type MessageNotification = {
  readonly id: string;
  readonly level: Level;
  readonly body: string;
};

/**
 * factory function to create one-shot notification
 * @param id ID of a notification
 * @param level level of a notification
 * @param message body of a message notification
 */
export const createMessage = function createMessage({
  id,
  level,
  message,
}: {
  id: string;
  level: Level;
  message: string;
}): MessageNotification {
  return { id, level, body: message };
};
