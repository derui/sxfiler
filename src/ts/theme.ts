import { createContext } from "preact";
import { useContext } from "preact/hooks";

// The context to get theme. The implementation of theme in this project is simple CSS class name for DOM.
export const ThemeContext = createContext<string | undefined>(undefined);

// too simple hook to use theme in container component
export const useTheme = () => {
  return useContext(ThemeContext);
};
