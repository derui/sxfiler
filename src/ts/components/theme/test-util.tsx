import * as React from "react";
import { Theme, ThemeProvider } from "./default-theme";

/**
   wrap element with default theme.
 */
export const wrap = function wrap(e: React.ReactElement) {
  return <ThemeProvider theme={Theme}>{e}</ThemeProvider>;
};
