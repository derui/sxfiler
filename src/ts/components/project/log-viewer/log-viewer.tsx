import * as React from "react";

import * as ListItem from "../../ui/list-item/list-item";
import * as List from "../../ui/list/list";
import { MessageNotification, Level } from "../../../domains/message-notification";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles: ClassNames = require("./log-viewer.module.scss");

export interface ClassNames {
  root?: string;
  item?: string;
  itemLevel?: string;
  itemMessage?: string;
}

export type Props = {
  entries: MessageNotification[];
  hidden: boolean;
};

/**
 * make item for a log entry
 */
const makeItem = (entry: MessageNotification) => {
  let level = "info";
  switch (entry.level) {
    case Level.Info:
      level = "info";
      break;
    case Level.Warning:
      level = "warning";
      break;
    case Level.Error:
      level = "error";
      break;
  }

  return (
    <ListItem.Component key={entry.id} className={styles.item}>
      <span className={styles.itemLevel} data-level={level}>
        [{level.toUpperCase()}]
      </span>
      <span className={styles.itemMessage}>{entry.body}</span>
    </ListItem.Component>
  );
};

export const Component: React.FC<Props> = ({ entries, hidden }) => {
  let components: React.ReactElement[] = [];
  if (!hidden) {
    components = entries.map(makeItem);
  }

  return (
    <List.Component className={styles.root} aria-hidden={hidden}>
      {components}
    </List.Component>
  );
};
