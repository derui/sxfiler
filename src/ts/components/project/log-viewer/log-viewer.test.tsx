import * as React from "react";
import renderer from "react-test-renderer";
import { wrap } from "@/components/theme/test-util";

import { Component as T } from "./log-viewer";
import { createMessage, Level } from "@/domains/message-notification";

describe("Project", () => {
  describe("Log Viewer", () => {
    it("should render empty", () => {
      const tree = renderer.create(wrap(<T entries={[]} hidden={false} />)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should render items each level", () => {
      const entries = [
        createMessage({ id: "id1", message: "info", level: Level.Info }),
        createMessage({ id: "id2", message: "error", level: Level.Error }),
        createMessage({ id: "id3", message: "warning", level: Level.Warning }),
      ];
      const tree = renderer.create(wrap(<T entries={entries} hidden={false} />)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should render empty when hidden it", () => {
      const entries = [
        createMessage({ id: "id1", message: "info", level: Level.Info }),
        createMessage({ id: "id2", message: "error", level: Level.Error }),
        createMessage({ id: "id3", message: "warning", level: Level.Error }),
      ];
      const tree = renderer.create(wrap(<T entries={entries} hidden={true} />)).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
