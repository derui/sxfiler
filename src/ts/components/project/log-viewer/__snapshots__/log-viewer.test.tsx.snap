// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Log Viewer should render empty 1`] = `
<div
  aria-hidden={false}
  className="root"
  role="list"
/>
`;

exports[`Project Log Viewer should render empty when hidden it 1`] = `
<div
  aria-hidden={true}
  className="root"
  role="list"
/>
`;

exports[`Project Log Viewer should render items each level 1`] = `
<div
  aria-hidden={false}
  className="root"
  role="list"
>
  <div
    aria-selected={false}
    className="item"
    role="listitem"
  >
    <span
      className="itemLevel"
      data-level="info"
    >
      INFO
    </span>
    <span
      className="itemMessage"
    >
      info
    </span>
  </div>
  <div
    aria-selected={false}
    className="item"
    role="listitem"
  >
    <span
      className="itemLevel"
      data-level="error"
    >
      ERROR
    </span>
    <span
      className="itemMessage"
    >
      error
    </span>
  </div>
  <div
    aria-selected={false}
    className="item"
    role="listitem"
  >
    <span
      className="itemLevel"
      data-level="warning"
    >
      WARNING
    </span>
    <span
      className="itemMessage"
    >
      warning
    </span>
  </div>
</div>
`;
